src/watchman_client.js

```diff
@@ -0,0 +1,376 @@
+    });
+};
+
+/**
+ * Remove the information about a specific WatchmanWatcher.
+ * Once done, if no watchers are left, clear the local vars,
+ * which will end the connection to the watchman.Client, too.
+ */
+WatchmanClient.prototype.closeWatcher = function(watchmanWatcher) {
+  let watcherInfos = Object.values(this._watcherMap);
+  let numWatchers = watcherInfos.length;
+
+  if (numWatchers > 0) {
+    let watcherInfo;
+    
+    for (let info of watcherInfos) {
+      if (info.watchmanWatcher === watchmanWatcher) {
+        watcherInfo = info;
+        break;
+      }
+    }
+
+    if (watcherInfo) {
+      delete this._watcherMap[watcherInfo.subscription];
+
+      numWatchers--;
+
+      if (numWatchers === 0) {
+        this._clearLocalVars();
+      }
+    }
+  }
+};
+
+/**
+ * Clear out local state at the beginning and if we end up
+ * getting disconnected and try to reconnect.
+ */
+WatchmanClient.prototype._clearLocalVars = function() {
+  if (this._client) {
+    this._client.removeAllListeners();
+    this._client.end();
+  }
+
+  this._client = null;
+  this._clientPromise = null;
+  this._wildmatch = false;
+  this._relative_root = false;
+  this._subscriptionId = 1;
+  this._watcherMap = {};
+  this._watchmanBinaryPath = null;
+};
+
+/**
+ * This singleton is the only object directly listening to our
+ * watchman.Client. Set up our listeners.
+ */
+WatchmanClient.prototype._setupClientListeners = function() {
+  this._client.on('subscription', this._onSubscription.bind(this));
+  this._client.on('error', this._onError.bind(this));
+  this._client.on('end', this._onEnd.bind(this));
+};
+
+WatchmanClient.prototype._getSubscription = function() {
+  let val = 'sane_' + this._subscriptionId++;
+  return val;
+};
+
+/**
+ * Create a new watcherInfo entry for the given watchmanWatcher and 
+ * initialize it.
+ */
+WatchmanClient.prototype._createWatcherInfo = function(watchmanWatcher) {
+  let watcherInfo = {
+    subscription: this._getSubscription(),
+    watchmanWatcher: watchmanWatcher,           
+    root: null,                      // set during 'watch' or 'watch-project'
+    relativePath: null,              // same
+    since: null,                     // set during 'clock'
+    options: null                    // created and set during 'subscribe'.
+  };
+
+  this._watcherMap[watcherInfo.subscription] = watcherInfo;
+
+  return watcherInfo;
+};
+
+/**
+ * Find an existing watcherInfo instance.
+ */
+WatchmanClient.prototype._getWatcherInfo = function(subscription) {
+  var watcherInfo = this._watcherMap[subscription];
+  return watcherInfo;
+}
+
+/**
+ * Given a watchmanWatcher and a root, issue the correct 'watch'
+ * or 'watch-project' command and handle it with the callback.
+ * Because we're operating in 'sane', we'll keep the results
+ * of the 'watch' or 'watch-project' here.
+ */
+WatchmanClient.prototype._watch = function(subscription, root) {
+  
+  let promise = new Promise((resolve, reject) => {
```

Comment:
i believe we can simply `return new Promise(...` here

Reply:
We can, yes. I sometimes do things this way so I can look at the value before returning it (i.e. having a breakpoint before the return), but I'll fix the various cases where I do this to just return the value w/o storing it in a local var first.

Reply:
Done.